### PR TITLE
Mutual friend

### PR DESCRIPTION
**Pull request by kbjude**
This pull request contains code that implements the mutual friendship to our app. It also has the tests(specs) for the same

### CODE REVIEW

File: app/models/friendship.rb [hunk 1]
```diff
@@ -1,4 +1,8 @@
 class Friendship < ApplicationRecord
   belongs_to :user
   belongs_to :friend, class_name: 'User'
+
+  def mutual
+    user.friends & friend.friends
+  end
```
Comment (majovanilla) [inline app/models/friendship.rb:7]: Nice start!! But the purpose of this milestone is to actually use the method!! In this case, the method is not actually creating the double friendship, but just making a collection of the user's friends and the friend's friends which will be unnecessary once you create the double row friendship.

Please check the comments I left in the main message.

**Review by majovanilla — CHANGES_REQUESTED** [review]
Hey @kbjude , 👋

## Good job on this project! 👍 You have a very complete set of associations! Now let's use them.

There are still some changes 👇 you need to implement to approve this request 🔍

### Required Changes: ✍️

- The method you created (mutual) only shows the user's friends and inverse friendship but doesn't create them.

- When friendship is confirmed, an additional row with inverse user-friend pair should be created. You can add a method on your User Model and call it when your friendship is confirmed.

- You have a method `pending_friends` on the User Model that is not necessary since you already declared the proper association. Instead of using that method, you can call the association whenever you need it (`user.pending_friends`)

- The same with the method `friends`. When you implement the double friendship, this method will not be necessary and you will need to call the `mutual` method you just created.

- You know can create a method on the User model to show the user's and their friend's posts. That way. you call this method on your timeline_posts method and you only show the user the posts they are allowed to see.

- From the previous milestone you forgot to add a User Show page. I should be able to click on the user's name and see their profile. Here, the above method comes handy, too. That way, a user can click on another user's profile without seeing their posts if they are not friends. Check the requirement [here](https://gitlab.com/microverse/guides/projects/requirements_screenshots/raw/master/images/rails/final_project/milestone5/user_can_see_invite_buttton.png).

- When I accept a friendship and I go to the All User menu, I see the button to send a friendship request, and that shouldn't happen since we are already friends!

- The above affects the timeline posts, too, since I'm unable to see my friend's posts. I think when you implement the required changes you will be able to fix this.

- After you implement the above, remember to add the necessary tests.

### Optional Changes:

If you want to make your project even better 🔥💥, I have a few suggestions you can implement.  

These are completely [Optional]: 🦔

- You should create a new menu called Notifications to show your pending and sent friend requests, instead of having them on the All Users menu (they serve different purposes) You could also make use of bootstrap badges to add a number to show the notifications you have!

## Please, make all REQUIRED changes and submit your request again! 😀

If you have any questions, feel free to send me a message through Slack (@majo)
I would be happy to help! 👩‍💻 


Happy coding 🤓

--Maria Reyes

**Comment by kbjude** [comment]
All feedback from the TSE implemented. A few of the things were unclear.

**Review by kobiyoyo — CHANGES_REQUESTED** [review]
## Hi @kbjude     
#### Great Job so far, I love every bit of this project from the `PR` down to the `Readme file`, let just fix few more things required. 
### General
- [ ] Anytime I accept a friend request I still get `Send Request`  button, lets fix that.
- [ ] Lets write unit and integration test for all components(Like, Comment & Post). 
- [ ] Kindly delete all commented codes.


### Make changes and submit another code review request.
*Happy coding* 
(Any questions Feel free to contact me on slack @kobiyoyo)

**Review by maelfosso — CHANGES_REQUESTED** [review]
# Changes required :red_circle: 

Hello,
Thank you for the improvements done.
However, there are some issues to fix before moving forward.

1. I created two account **Bob Bar** and **Alice Bar**. There are now friends
Screenshot at 2020-06-29 14:38:39

But here is the table content
```rb
2.6.3 :001 > Friendship.all
  Friendship Load (0.6ms)  SELECT  "friendships".* FROM "friendships" LIMIT $1  [["LIMIT", 11]]
 => #<ActiveRecord::Relation [#<Friendship id: 1, user_id: 2, friend_id: 1, confirmed: true, created_at: "2020-06-29 13:35:10", updated_at: "2020-06-29 13:36:15">]> 
2.6.3 :002 > 
```

As you can notice, there is only one line instead of two as requested

2- Also, the two users are now friends and there is something weird.
Look at the timeline from **Alice Bar** user
Screenshot at 2020-06-29 14:41:26

Now, look at the timeline from **Bob Bar** view
Screenshot at 2020-06-29 14:42:03

It's different which is not good.
The errors up and this one are related.

3- I got some errors from **Rspec**
```rb
  1) First page Before accesing the post page redirects to the home page with 'sign up'
     Got 0 failures and 2 other errors:

     1.1) Failure/Error: visit root_path
          
          LoadError:
            Capybara's selenium driver is unable to load `selenium-webdriver`, please install the gem and add `gem 'selenium-webdriver'` to your Gemfile if you are using bundler.
          
          
          
          # /home/maelfosso/.rvm/gems/ruby-2.6.3/gems/capybara-3.32.2/lib/capybara/selenium/driver.rb:30:in `rescue in load_selenium'
          # /home/maelfosso/.rvm/gems/ruby-2.6.3/gems/capybara-3.32.2/lib/capybara/selenium/driver.rb:18:in `load_selenium'
          # /home/maelfosso/.rvm/gems/ruby-2.6.3/gems/capybara-3.32.2/lib/capybara/selenium/driver.rb:61:in `initialize'
          # /home/maelfosso/.rvm/gems/ruby-2.6.3/gems/capybara-3.32.2/lib/capybara/session.rb:104:in `driver'
          # /home/maelfosso/.rvm/gems/ruby-2.6.3/gems/capybara-3.32.2/lib/capybara/session.rb:90:in `initialize'
          # /home/maelfosso/.rvm/gems/ruby-2.6.3/gems/capybara-3.32.2/lib/capybara.rb:415:in `new'
          # /home/maelfosso/.rvm/gems/ruby-2.6.3/gems/capybara-3.32.2/lib/capybara.rb:415:in `block in session_pool'
          # /home/maelfosso/.rvm/gems/ruby-2.6.3/gems/capybara-3.32.2/lib/capybara.rb:312:in `current_session'
          # /home/maelfosso/.rvm/gems/ruby-2.6.3/gems/capybara-3.32.2/lib/capybara/dsl.rb:46:in `page'
          # /home/maelfosso/.rvm/gems/ruby-2.6.3/gems/capybara-3.32.2/lib/capybara/dsl.rb:58:in `block (2 levels) in <module:DSL>'
          # ./spec/views/index_spec.rb:4:in `block (3 levels) in <main>'
          # ------------------
          # --- Caused by: ---
          # LoadError:
          #   cannot load such file -- selenium-webdriver
          #   /home/maelfosso/.rvm/gems/ruby-2.6.3/gems/bootsnap-1.4.6/lib/bootsnap/load_path_cache/core_ext/kernel_require.rb:34:in `require'

     1.2) Failure/Error: raise LoadError, "Capybara's selenium driver is unable to load `selenium-webdriver`, please install the gem and add `gem 'selenium-webdriver'` to your Gemfile if you are using bundler."
          
          LoadError:
            Capybara's selenium driver is unable to load `selenium-webdriver`, please install the gem and add `gem 'selenium-webdriver'` to your Gemfile if you are using bundler.
          
          
          
          # /home/maelfosso/.rvm/gems/ruby-2.6.3/gems/capybara-3.32.2/lib/capybara/selenium/driver.rb:30:in `rescue in load_selenium'
          # /home/maelfosso/.rvm/gems/ruby-2.6.3/gems/capybara-3.32.2/lib/capybara/selenium/driver.rb:18:in `load_selenium'
          # /home/maelfosso/.rvm/gems/ruby-2.6.3/gems/capybara-3.32.2/lib/capybara/selenium/driver.rb:61:in `initialize'
          # /home/maelfosso/.rvm/gems/ruby-2.6.3/gems/capybara-3.32.2/lib/capybara/session.rb:104:in `driver'
          # /home/maelfosso/.rvm/gems/ruby-2.6.3/gems/capybara-3.32.2/lib/capybara/session.rb:90:in `initialize'
          # /home/maelfosso/.rvm/gems/ruby-2.6.3/gems/capybara-3.32.2/lib/capybara.rb:415:in `new'
          # /home/maelfosso/.rvm/gems/ruby-2.6.3/gems/capybara-3.32.2/lib/capybara.rb:415:in `block in session_pool'
          # /home/maelfosso/.rvm/gems/ruby-2.6.3/gems/capybara-3.32.2/lib/capybara.rb:312:in `current_session'
          # /home/maelfosso/.rvm/gems/ruby-2.6.3/gems/capybara-3.32.2/lib/capybara/dsl.rb:46:in `page'
          # ------------------
          # --- Caused by: ---
          # LoadError:
          #   cannot load such file -- selenium-webdriver
          #   /home/maelfosso/.rvm/gems/ruby-2.6.3/gems/bootsnap-1.4.6/lib/bootsnap/load_path_cache/core_ext/kernel_require.rb:34:in `require'

  2) User Friendship Friendhip not created when there is no second user
     Failure/Error: friends.include?(user) || confirmed_inverse_friendships.map(&:user_id).include?(user.id)
     
     NoMethodError:
       undefined method `id' for "user2":String
     # ./app/models/user.rb:47:in `friend?'
     # ./spec/models/user_spec.rb:7:in `block (3 levels) in <main>'

Finished in 0.72344 seconds (files took 1.58 seconds to load)
8 examples, 2 failures

Failed examples:

rspec ./spec/views/index_spec.rb:3 # First page Before accesing the post page redirects to the home page with 'sign up'
rspec ./spec/models/user_spec.rb:5 # User Friendship Friendhip not created when there is no second user

Randomized with seed 47337
```
Also, being sure that the integration tests take in consideration the fact that the Friendship concerns two lines now.

Kindly make the changes and submit a new code review request
Thanks

Mael FOSSO

**Comment by kbjude** [comment]
Fixed feedback from the TSE

**Comment by uryelah** [comment]
Hi **Jude Kajura Balinda**, I'm Sarah your reviewer for this PR 👩‍🚀 

Good job going through the reviews and fixing all the pointed out issues so far!
Your project is good enough to pass now, congrats!

-------

## **Your project is approved! 🎊 🥅 💛**

**You may close this Pull Request and merge the changes**


---------

    
    
**Keep being an awesome dev and good luck with your next project!**

---------

Cheers and Happy coding!👏👏👏

Code review by **Sarah Uryelah Silva Chamorro**

(Feel free to contact me on slack by the handle of  **sarah** if you have any additional question)

**Pull request merged**